tests/test_non_linearity_correction.py

```diff
@@ -121,28 +121,46 @@ def test_non_linearity_correction():
     
     Ported from II&T Pipeline
     """
+
+    #Create a mock dataset because it is a required input when creating a NonLinearityCalibration
+    dummy_dataset = mocks.create_prescan_files()
+
+    # Make a non-linearity correction calibration file
+    input_non_linearity_filename = "nonlin_table_TVAC.txt"
+    input_non_linearity_path = os.path.join(os.path.dirname(__file__), "test_data", input_non_linearity_filename)
+    test_non_linearity_filename = input_non_linearity_filename.split(".")[0] + ".fits"
+    test_non_linearity_path = os.path.join(os.path.dirname(__file__), "test_data", test_non_linearity_filename)
+    tvac_nonlin_data = np.genfromtxt(input_non_linearity_path, delimiter=",")
+
+
+    pri_hdr, ext_hdr = mocks.create_default_headers()
+    non_linearity_correction = data.NonLinearityCalibration(tvac_nonlin_data,pri_hdr=pri_hdr,ext_hdr=ext_hdr,input_dataset = dummy_dataset)
+    non_linearity_correction.save(filename = test_non_linearity_path)
+    # COMMENT CSV data saved as FITS file                                             
+    # HISTORY nonlin_sample.CSV from IIT data converted to FITS  
+
+    # import IPython; IPython.embed()
+
     ###### create a simulated dataset that is non-linear
     # check that simulated data folder exists, and create if not
     datadir = os.path.join(os.path.dirname(__file__), "simdata")
     if not os.path.exists(datadir):
         os.mkdir(datadir)
-    
-    nonlin_fits_filepath = os.path.join(os.path.dirname(__file__), "test_data", "nonlin_sample.fits")
 
     emgain = 2000
-    mocks.create_nonlinear_dataset(nonlin_fits_filepath, filedir=datadir,em_gain=emgain)
+    mocks.create_nonlinear_dataset(test_non_linearity_path, filedir=datadir,em_gain=emgain)
 
     ####### open up the files
     sim_data_filenames = glob.glob(os.path.join(datadir, "simcal_nonlin*.fits"))
     nonlinear_dataset = data.Dataset(sim_data_filenames)
     assert len(nonlinear_dataset) == 2
 
     ######## perform non-linearity correction
-    non_linearity_correction = data.NonLinearityCalibration(os.path.join(os.path.dirname(__file__),"test_data",'nonlin_sample.fits'))
+    non_linearity_correction = data.NonLinearityCalibration(test_non_linearity_path)
     linear_dataset = l1_to_l2a.correct_nonlinearity(nonlinear_dataset, non_linearity_correction)
 
     #The data was generated with a ramp in the x-direction going from 10 to 65536
```

Comment:
This is a good question and something that probably needs a bit more thinking. Without changing this, one of the tests was getting tripped in the very low SNR limit. Think mean signal mean of 1, with noise of +/-1 . Why is it tripping with the TVAC non-linearity curves, but not the testing ones? I'm not sure. This might mean that we want to dig into this a bit more and put off this PR for now. Thoughts?

Reply:
The TVAC table starts out at 800 counts, whereas the sample nonlin table starts at 1 count. Could be an issue with extrapolating the table out of bounds?

Reply:
Does it work if you start from 800?

Reply:
Oh, good catch. That's probably why it didn't work. 

However, annoyingly it doesn't work if we start at 800, which is confusing.

Reply:
Alright, I've figured this out. It has to do with the fact that the corrections are calculated based on a dataset. When generating a mock dataset we were faking non-linearity based on a perfect input dataset by dividing through by the corrections (rather than multiplying, as we do on the real data). When calculating the gains to apply to the mock data, as input I was using the perfect (linear) input data. When calculating the gains to apply to the real data, as input I was using the non-linear data. This meant that the gains were slightly different between the input mock data and the corrected data, which resulted in the comparison failing in some cases. I've rectified that applying an iterative approach to the mock data (just one iteration) that helps better match the gains applied to the mock data and the input data. The data now agrees to within 1%, as was the previous tests. 

(I also updated the range of testing values to be starting from 800, which is where the TVAC non-linearity table starts - good catch Jason!).